### ISSUE TITLE
Support arbitrary extensions

### ISSUE THREAD

**Issue by smichel17** [jmhobbs/jsTodoTxt#10]
Rather than just having support for hidden and due extensions and adding additional extensions as they are requested, we could change it so the data members of an item are:
```
text      // The core text of the item
priority  // The priority level of the item as string [A-Z]
complete  // Boolean
completed // Date completed
date      // Date attached to object, typically with a comlete item
contexts  // Array, contexts for the object, e.g. @Home == "Home"
projects  // Array, projects fot the object, e.g. +Chores == "Chores"
extensions //Array using key:value pairs, where the key is the extension and the value is an array of values for that extension
```

So for example, `example task dep:1 dep:2 h:1` would parse to
```
{ text: "third task"
  extensions: { dep: [ '1', '2' ]
                h: [ '1' ]
              }
}
```

It allows anyone using the jsTodoTxt to implement their own custom extensions without building in behavior that breaks the todo.txt standard (or having to do other people's work for them, if they want to use this to parse tasks and support their own custom extension).

**Comment by jmhobbs** [comment]
That seems reasonable, I'm looking for the counter case here though.

Is `<key>:<value>` the only way we would see it being extended?  Perhaps extensions might contain more logic than that though, is it reasonable to pull that logic up a level?

I wonder if we should implement an extension system that allows for deeper parsing/composition hooks, of which a key:value extension could be one.

Thoughts @bartlibert?

**Comment by smichel17** [comment]
What kind of deeper parsing are you thinking of?

`<key>:<value>` is the [format for add-ons](https://github.com/ginatrapani/todo.txt-cli/wiki/The-Todo.txt-Format#add-on-file-format-definitions) in the todo.txt spec. IMO, this library should stick to a barebones implementation of that spec, and leave any deeper parsing (eg, implementing an add-on) to another library (which might also be authored by you).

**Comment by smichel17** [comment]
If I were going to implement from scratch, I think I would use this data structure, which I will describe in examples instead of words because js arrays are "wonderful".
```
var task = new TodoTxtItem( "(A) 2016-09-27 Log issue on @github for +jsTodoTxt @computer due:2016-09-28" h:1);

console.log(task[0].text);
// "(A)"
console.log(task[1].type);
// "date"
console.log(task[5].value);
// "github"
console.log(task[10].type); // No thought went into choosing a symbol
// "*h"

console.log(task.filter());
// "(A) 2016-09-27 Log issue on @github for +jsTodoTxt @computer due:2016-09-28"
console.log(task.filter( ["priority", "*due"] );
// "2016-09-27 Log issue on @github for +jsTodoTxt @computer h:1"
console.log(task.parts( ["priority", "text", "context", "project", "*h"] );
// "2016-09-27 Log issue on @github for +jsTodoTxt @computer h:1"

console.log(task.context[0]); // Internally, just store [5] and this will be a function call
// "github"
console.log(task.contexts; // Internally, [5,8]
// ["github", "computer"]
console.log(task["*due"])); // Internally, [9]
// ["2016-09-28"]
console.log(task["*t"]); // Check if the "t" extension is in use
// undefined
```

**Comment by jmhobbs** [comment]
> What kind of deeper parsing are you thinking of?

No clue!  Just wondering if it was a thing.

> <key>:<value> is the format for add-ons in the todo.txt spec. I

In that case, yes, we should support it explicitly.

**Comment by jmhobbs** [comment]
I think we should move the extension parsing into the main body, but I would still like to include the possibility of explicit extensions including their own logic on top, i.e. the `DueExtension` parsing and validating the date.

**Comment by smichel17** [comment]
I agree that it would be nice to have support for extensions like hidden, due, threshold... but why not provide those in another library that wraps this one? eg, `jsTodoTxtExt`

Or we could provide those as utility methods that aren't baked into the object, like render.
